File: projeto/python/Servidor_sync.py
import zmq
import time
import threading
import uuid
import os
from datetime import datetime
from LogicalClock import LogicalClock

canais = []
users = []
relogio_logico = LogicalClock()
server_name = None
coordenador = None
mensagens_desde_sync = 0
SYNC_INTERVAL = 15

ALL_SERVERS = ["servidor-1", "servidor-2"]

def formatar_hora(timestamp):
    dt = datetime.fromtimestamp(timestamp / 1000.0)
    return dt.strftime("%H:%M:%S.%f")[:-3]

def log(msg):
    print(f"[{datetime.now()}] [{server_name}] {msg}")

def get_channel_file():
    return os.path.join(os.getcwd(), f"canais_{server_name}.txt")

def load_channels():
    path = get_channel_file()
    if os.path.exists(path):
        with open(path, "r") as f:
            loaded = [line.strip() for line in f if line.strip()]
        canais.clear()
        canais.extend(loaded)
        log(f"Canais carregados do arquivo: {canais}")
    else:
        log("Nenhum arquivo de canais local encontrado.")

def save_channels():
    path = get_channel_file()
    with open(path, "w") as f:
        for canal in sorted(set(canais)):
            f.write(canal + "\n")

def merge_channels(new_channels):
    changed = False
    for canal in new_channels:
        canal = canal.strip()
        if canal and canal not in canais:
            canais.append(canal)
            changed = True
    if changed:
        save_channels()
        log(f"Sincronizei canais locais: {canais}")
    return changed

def sync_channels_with_reference(ref):
    if canais:
        canais_str = ",".join(canais)
        ref.send_string(f"SYNC_CHANNELS {server_name} {canais_str}")
    else:
        ref.send_string("GET_CHANNELS")

    response = ref.recv_string()
    if response and response != "nenhum canal":
        merge_channels(response.split("\n"))
    else:
        log("Referência não retornou canais ou lista vazia.")

def main():
    global server_name, coordenador, mensagens_desde_sync

    hostname = os.getenv("HOSTNAME")
    if hostname and (hostname.startswith("servidor-") or hostname.startswith("servidor_")):
        server_name = hostname
    else:
        server_name = f"servidor-{str(uuid.uuid4())[:8]}"

    log("Iniciando servidor...")

    context = zmq.Context()

    rep = context.socket(zmq.REP)
    rep.connect("tcp://broker:5556")

    pub = context.socket(zmq.PUB)
    pub.connect("tcp://proxy:5557")

    ref = context.socket(zmq.REQ)
    ref.connect("tcp://reference:6000")

    load_channels()

    ref.send_string(f"REGISTER {server_name}")
    rank = ref.recv_string()
    log(f"Registrado: {rank}")

    try:
        sync_channels_with_reference(ref)
    except Exception as e:
        log(f"Falha na sincronização inicial de canais: {e}")

    time.sleep(5)

    # ELEIÇÃO
    ordenados = sorted(ALL_SERVERS)
    coordenador = ordenados[0]
    sou_coord = coordenador == server_name

    log(f"SERVIDORES: {ALL_SERVERS}")
    log(f"COORDENADOR ELEITO: {coordenador}")
    log(f"EU SOU COORDENADOR? {sou_coord}")

    pub.send_string(f"{relogio_logico.increment()}|servers {coordenador}")

    coord_socket = None
    if sou_coord:
        coord_socket = context.socket(zmq.REP)
        coord_socket.bind("tcp://*:5998")
        log("COORDENADOR: Ouvindo requisições de hora na porta 5998")

    eleicao_socket = context.socket(zmq.REP)
    eleicao_socket.bind("tcp://*:5999")

    def eleicao_thread():
        while True:
            msg = eleicao_socket.recv_string()
            if msg:
                eleicao_socket.send_string("OK")

    threading.Thread(target=eleicao_thread, daemon=True).start()

    if sou_coord:
        def coord_thread():
            while True:
                req = coord_socket.recv_string()
                if req == "REQ_HORA":
                    hora = int(time.time() * 1000)
                    coord_socket.send_string(f"REP_HORA|{hora}")
                    log(f"Forneci hora: {formatar_hora(hora)}")

        threading.Thread(target=coord_thread, daemon=True).start()

    log("Servidor pronto para processar mensagens")

    while True:
        mensagem = rep.recv_string()
        partes_clock = mensagem.split("|", 1)
        clock_recebido = int(partes_clock[0])
        comando = partes_clock[1]

        relogio_logico.update(clock_recebido)

        tokens = comando.split(" ", 2)
        cmd = tokens[0]
        resposta = ""

        if cmd == "logar":
            resposta = "usuario logado"
        elif cmd == "adiciona":
            canal = tokens[1]
            if canal not in canais:
                canais.append(canal)
                save_channels()
                try:
                    ref.send_string(f"CHANNEL_UPDATE {server_name} {canal}")
                    ack = ref.recv_string()
                    log(f"Atualizei referência de canais: {ack}")
                except Exception as e:
                    log(f"Falha ao atualizar referência de canais: {e}")
            resposta = "\n".join(canais)
        elif cmd == "lista":
            resposta = "nenhum canal" if not canais else "\n".join(canais)
        elif cmd == "publica":
            canal = tokens[1]
            msg = tokens[2]
            timestamp = str(datetime.now())
            pub.send_string(f"{relogio_logico.increment()}|{canal} {timestamp} {msg}")
            log(f"Publicado em {canal}: {msg}")
            resposta = "OK"

        rep.send_string(f"{relogio_logico.increment()}|{resposta}")

        mensagens_desde_sync += 1

        if mensagens_desde_sync % 10 == 0:
            try:
                ref.send_string(f"HEARTBEAT {server_name}")
                hb_resp = ref.recv_string()
                log(f"Heartbeat: {hb_resp}")
            except Exception as e:
                log(f"Falha no heartbeat: {e}")
            log(f"Coordenador atual: {coordenador}")

        if mensagens_desde_sync % SYNC_INTERVAL == 0:
            log("INICIANDO SINCRONIZAÇÃO DE CANAIS COM REFERÊNCIA...")
            try:
                sync_channels_with_reference(ref)
            except Exception as e:
                log(f"Falha de sincronização de canais: {e}")

            if not sou_coord:
                log("INICIANDO SINCRONIZAÇÃO BERKELEY...")
                sync_client = context.socket(zmq.REQ)
                sync_client.setsockopt(zmq.RCVTIMEO, 3000)

                try:
                    sync_client.connect(f"tcp://{coordenador}:5998")

                    t1 = int(time.time() * 1000)
                    sync_client.send_string("REQ_HORA")
                    resposta_hora = sync_client.recv_string()
                    t2 = int(time.time() * 1000)

                    if resposta_hora and resposta_hora.startswith("REP_HORA|"):
                        hora_coord = int(resposta_hora.split("|")[1])
                        rtt = t2 - t1
                        hora_ajustada = hora_coord + (rtt // 2)
                        diferenca = hora_ajustada - int(time.time() * 1000)

                        log(f"Hora do coordenador: {formatar_hora(hora_coord)}")
                        log(f"RTT: {rtt}ms")
                        log(f"Hora ajustada:     {formatar_hora(hora_ajustada)}")
                        log(f"Diferença: {diferenca:+}ms")
                        log(f"Relógio lógico: {relogio_logico.get_time()}")

                except Exception as e:
                    log(f"ERRO: Coordenador {coordenador} falhou! {e}")
                    log("Iniciando nova eleição...")

                    idx = ALL_SERVERS.index(coordenador)
                    if idx >= 0 and idx + 1 < len(ALL_SERVERS):
                        coordenador = ALL_SERVERS[idx + 1]
                    else:
                        coordenador = ALL_SERVERS[0]
                    log(f"NOVO COORDENADOR: {coordenador}")
                    pub.send_string(f"{relogio_logico.increment()}|servers {coordenador}")

                sync_client.close()

if __name__ == "__main__":
    main()

File: projeto/java/Servidor_sync.java
import org.zeromq.ZMQ;
import java.util.*;
import java.time.LocalDateTime;
import java.time.Instant;
import java.time.format.DateTimeFormatter;
import java.io.*;
import java.nio.file.*;

public class Servidor_sync {

    private static List<String> canais = new ArrayList<>();
    private static List<String> users = new ArrayList<>();
    private static LogicalClock relogioLogico = new LogicalClock();
    private static String serverName;
    private static String coordenador = null;
    private static int mensagensDesdeSync = 0;
    private static final int SYNC_INTERVAL = 15;

    private static final List<String> ALL_SERVERS = Arrays.asList(
        "servidor-1", "servidor-2"
    );

    private static final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern("HH:mm:ss.SSS");

    public static String formatarHora(long timestamp) {
        return Instant.ofEpochMilli(timestamp)
                .atZone(java.time.ZoneId.systemDefault())
                .format(FORMATTER);
    }

    public static void log(String msg) {
        System.out.println("[" + LocalDateTime.now() + "] [" + serverName + "] " + msg);
    }

    private static Path getChannelFile() {
        return Paths.get(System.getProperty("user.dir"), "canais_" + serverName + ".txt");
    }

    private static void loadChannels() {
        Path path = getChannelFile();
        if (Files.exists(path)) {
            try {
                List<String> loaded = Files.readAllLines(path);
                canais.clear();
                for (String line : loaded) {
                    line = line.trim();
                    if (!line.isEmpty()) {
                        canais.add(line);
                    }
                }
                log("Canais carregados do arquivo: " + canais);
            } catch (IOException e) {
                log("Erro ao carregar canais locais: " + e.getMessage());
            }
        } else {
            log("Nenhum arquivo de canais local encontrado.");
        }
    }

    private static void saveChannels() {
        Path path = getChannelFile();
        try (BufferedWriter writer = Files.newBufferedWriter(path)) {
            Set<String> unique = new TreeSet<>(canais);
            for (String canal : unique) {
                writer.write(canal);
                writer.newLine();
            }
        } catch (IOException e) {
            log("Erro ao salvar canais locais: " + e.getMessage());
        }
    }

    private static boolean mergeChannels(Collection<String> newChannels) {
        boolean changed = false;
        for (String canal : newChannels) {
            canal = canal.trim();
            if (!canal.isEmpty() && !canais.contains(canal)) {
                canais.add(canal);
                changed = true;
            }
        }
        if (changed) {
            saveChannels();
            log("Sincronizei canais locais: " + canais);
        }
        return changed;
    }

    private static void syncChannelsWithReference(ZMQ.Socket ref) {
        if (!canais.isEmpty()) {
            String canaisStr = String.join(",", canais);
            ref.send("SYNC_CHANNELS " + serverName + " " + canaisStr);
        } else {
            ref.send("GET_CHANNELS");
        }

        String response = ref.recvStr();
        if (response != null && !response.equals("nenhum canal") && !response.trim().isEmpty()) {
            mergeChannels(Arrays.asList(response.split("\n")));
        } else {
            log("Referência não retornou canais ou lista vazia.");
        }
    }

    public static void main(String[] args) throws Exception {

        String hostname = System.getenv("HOSTNAME");
        if (hostname != null && (hostname.startsWith("servidor-") || hostname.startsWith("servidor_"))) {
            serverName = hostname;
        } else {
            serverName = "servidor-" + UUID.randomUUID().toString().substring(0, 8);
        }

        log("Iniciando servidor...");

        ZMQ.Context context = ZMQ.context(1);

        ZMQ.Socket rep = context.socket(ZMQ.REP);
        rep.connect("tcp://broker:5556");

        ZMQ.Socket pub = context.socket(ZMQ.PUB);
        pub.connect("tcp://proxy:5557");

        ZMQ.Socket ref = context.socket(ZMQ.REQ);
        ref.connect("tcp://reference:6000");
        ref.setReceiveTimeOut(3000);

        loadChannels();

        ref.send("REGISTER " + serverName);
        String rank = ref.recvStr();
        log("Registrado: " + rank);

        try {
            syncChannelsWithReference(ref);
        } catch (Exception e) {
            log("Falha na sincronização inicial de canais: " + e.getMessage());
        }

        Thread.sleep(5000);

        List<String> ordenados = new ArrayList<>(ALL_SERVERS);
        Collections.sort(ordenados);
        coordenador = ordenados.get(0);
        final boolean souCoord = coordenador.equals(serverName);

        log("SERVIDORES: " + ALL_SERVERS);
        log("COORDENADOR ELEITO: " + coordenador);
        log("EU SOU COORDENADOR? " + souCoord);

        pub.send(relogioLogico.increment() + "|servers " + coordenador);

        final ZMQ.Socket coordSocket;
        if (souCoord) {
            coordSocket = context.socket(ZMQ.REP);
            coordSocket.bind("tcp://*:5998");
            log("COORDENADOR: Ouvindo requisições de hora na porta 5998");
        } else {
            coordSocket = null;
        }

        final ZMQ.Socket eleicaoSocket = context.socket(ZMQ.REP);
        eleicaoSocket.bind("tcp://*:5999");

        new Thread(() -> {
            while (true) {
                String msg = eleicaoSocket.recvStr();
                if (msg != null) {
                    eleicaoSocket.send("OK");
                }
            }
        }).start();

        if (souCoord) {
            new Thread(() -> {
                while (true) {
                    String req = coordSocket.recvStr();
                    if (req != null && req.equals("REQ_HORA")) {
                        long hora = System.currentTimeMillis();
                        coordSocket.send("REP_HORA|" + hora);
                        log("Forneci hora: " + formatarHora(hora));
                    }
                }
            }).start();
        }

        log("Servidor pronto para processar mensagens");

        while (true) {
            String mensagem = rep.recvStr();
            String[] partesClock = mensagem.split("\\|", 2);
            int clockRecebido = Integer.parseInt(partesClock[0]);
            String comando = partesClock[1];

            relogioLogico.update(clockRecebido);

            String[] tokens = comando.split(" ", 3);
            String cmd = tokens[0];
            String resposta = "";

            if (cmd.equals("logar")) {
                resposta = "usuario logado";
            } else if (cmd.equals("adiciona")) {
                String canal = tokens[1];
                if (!canais.contains(canal)) {
                    canais.add(canal);
                    saveChannels();
                    try {
                        ref.send("CHANNEL_UPDATE " + serverName + " " + canal);
                        String ack = ref.recvStr();
                        log("Atualizei referência de canais: " + ack);
                    } catch (Exception e) {
                        log("Falha ao atualizar referência de canais: " + e.getMessage());
                    }
                }
                resposta = String.join("\n", canais);
            } else if (cmd.equals("lista")) {
                resposta = canais.isEmpty() ? "nenhum canal" : String.join("\n", canais);
            } else if (cmd.equals("publica")) {
                String canal = tokens[1];
                String msg = tokens[2];
                String timestamp = LocalDateTime.now().toString();
                pub.send(relogioLogico.increment() + "|" + canal + " " + timestamp + " " + msg);
                log("Publicado em " + canal + ": " + msg);
                resposta = "OK";
            }

            rep.send(relogioLogico.increment() + "|" + resposta);

            mensagensDesdeSync++;

            if (mensagensDesdeSync % 10 == 0) {
                try {
                    ref.send("HEARTBEAT " + serverName);
                    String hbResp = ref.recvStr();
                    log("Heartbeat: " + hbResp);
                } catch (Exception e) {
                    log("Falha no heartbeat: " + e.getMessage());
                }
                log("Coordenador atual: " + coordenador);
            }

            if (mensagensDesdeSync % SYNC_INTERVAL == 0) {
                log("INICIANDO SINCRONIZAÇÃO DE CANAIS COM REFERÊNCIA...");
                try {
                    syncChannelsWithReference(ref);
                } catch (Exception e) {
                    log("Falha de sincronização de canais: " + e.getMessage());
                }

                if (!souCoord) {
                    log("INICIANDO SINCRONIZAÇÃO BERKELEY...");

                    ZMQ.Socket syncClient = context.socket(ZMQ.REQ);
                    syncClient.setReceiveTimeOut(3000);

                    try {
                        syncClient.connect("tcp://" + coordenador + ":5998");

                        long t1 = System.currentTimeMillis();
                        syncClient.send("REQ_HORA");
                        String respostaHora = syncClient.recvStr();
                        long t2 = System.currentTimeMillis();

                        if (respostaHora != null && respostaHora.startsWith("REP_HORA|")) {
                            long horaCoord = Long.parseLong(respostaHora.split("\\|")[1]);
                            long rtt = t2 - t1;
                            long horaAjustada = horaCoord + (rtt / 2);
                            long diferenca = horaAjustada - System.currentTimeMillis();

                            log("Hora do coordenador: " + formatarHora(horaCoord));
                            log("RTT: " + rtt + "ms");
                            log("Hora ajustada:     " + formatarHora(horaAjustada));
                            log("Diferença: " + (diferenca >= 0 ? "+" : "") + diferenca + "ms");
                            log("Relógio lógico: " + relogioLogico.getTime());
                        }
                    } catch (Exception e) {
                        log("ERRO: Coordenador " + coordenador + " falhou! " + e.getMessage());
                        log("Iniciando nova eleição...");

                        int idx = ALL_SERVERS.indexOf(coordenador);
                        if (idx >= 0 && idx + 1 < ALL_SERVERS.size()) {
                            coordenador = ALL_SERVERS.get(idx + 1);
                        } else {
                            coordenador = ALL_SERVERS.get(0);
                        }
                        log("NOVO COORDENADOR: " + coordenador);
                        pub.send(relogioLogico.increment() + "|servers " + coordenador);
                    } finally {
                        syncClient.close();
                    }
                }
            }
        }
    }
}
